src/moi.jl

```diff
@@ -406,13 +413,186 @@ function build_bilevel(
+                end
+            else  # AB_N is not empty
+                @debug("set AB_N is NOT empty")
+                
+                # TODO input flag for checking for blocks? (to save time)
+                num_blocks, rows, cols = find_blocks(V, U)
+
+                conditions_passed = Bool[]
+                nrows, ncols = size(V)
+                for n in 1:num_blocks
+                    # TODO can skip blocks that are not linked to bilinear terms?
+                    Vblock = spzeros(nrows, ncols)
+                    Vblock[rows[n],:] = V[rows[n],:]
+                    check = check_non_empty_AB_N_conditions(
+                        intersect(J_U, rows[n]), U, intersect(N_U, cols[n]), intersect(A_N, cols[n]), B, Vblock, 
+                        lower_primal_var_to_lower_con, upper_var_to_lower_ctr,
+                        bilinear_upper_dual_to_quad_term, 
+                        bilinear_upper_dual_to_lower_primal)
+                    push!(conditions_passed, check)
+                end
+                # recover memory
+                rows, cols = nothing, nothing
+                @debug("Done looping over V blocks")
+                
+                if all(conditions_passed)
+                    linearizations = linear_terms_for_non_empty_AB(
+                        lower,
+                        upper_var_to_lower_ctr,
+                        bilinear_upper_dual_to_lower_primal,
+                        V,
+                        w,
+                        A_N,
+                        bilinear_upper_dual_to_quad_term,
+                        lower_obj_terms,
+                        lower_to_m_idxmap,
+                        lower_primal_dual_map,
+                        lower_dual_idxmap
+                    )
+                    if isnothing(linearizations)
+                        @warn("Unable to linearize bilinear terms due to underdertermined system of equations. Skipping linearization process.")
+                    end
+                else
+                    @warn("Required conditions for linearization not met. Skipping linearization process.")
+                end
+
+            end
+
+            # LINEARIZATION
+            # TODO switch signs with MAX sense?
+
+            if !(isnothing(linearizations))
+                # set m's objective by replacing quadratic terms with linearizations
+                mobj = deepcopy(m_objective)
+                quadratic_terms = mobj.quadratic_terms  # TODO keep quadratic terms that have not been linearized
+                c = mobj.constant
+                affine_terms = mobj.affine_terms
+                cvb = collect(values(bilinear_upper_quad_term_to_m_quad_term))
+                new_objective = deepcopy(m_objective)
+                if Set(cvb) == Set(quadratic_terms)
+                    @info("Replacing bilinear lower dual * lower primal terms in upper objective with linear terms.")
```

Comment:
`@debug`

Reply:
done in bc4539998c421fec88da7461efe8ebcdbdb6d0b9